js-pkg/client/src/provider.ts

```diff
@@ -469,4 +490,138 @@ export class YSweetProvider extends Observable<string> {
+    this.onFailureHandlers.push(handler)
+    return () => {
+      this.onFailureHandlers = this.onFailureHandlers.filter((h) => h !== handler)
+    }
+  }
+}
+
+async function getClientToken(authEndpoint: AuthEndpoint, roomname: string): Promise<ClientToken> {
+  if (typeof authEndpoint === 'function') {
+    return await authEndpoint()
+  }
+  const body = JSON.stringify({ docId: roomname })
+  const res = await fetch(authEndpoint, {
+    method: 'POST',
+    body,
+    headers: { 'Content-Type': 'application/json' },
+  })
+  // TODO: handle errors
+  const clientToken = await res.json()
+  // TODO: check that clientToken.docId === this.roomname
+  return clientToken
+}
+
+function updateProviderParams(
+  providerParams: YSweetProviderParams,
+  clientToken: ClientToken,
+): YSweetProviderParams {
+  // if the clientToken has a token, add it to the provider params
+  const connectParams = providerParams.params ? { ...providerParams.params } : {}
+  if (clientToken.token) connectParams.token = clientToken.token
+  return { ...providerParams, params: connectParams }
+}
+
+export async function ySweetProviderWrapper(
+  authEndpoint: AuthEndpoint,
+  roomname: string,
+  doc: Y.Doc,
+  providerParams: YSweetProviderParams = {},
+): Promise<YSweetProviderWithClientToken> {
+  const observable = providerParams.observable ?? new Observable<string>()
+  providerParams = { ...providerParams, observable }
+
+  let _clientToken = await getClientToken(authEndpoint, roomname)
+  let _provider = new YSweetProvider(_clientToken.url, roomname, doc, {
+    ...updateProviderParams(providerParams, _clientToken),
+  })
+  _provider.addOnFailureHandler(recreateProvider)
+
+  async function recreateProvider() {
+    _clientToken = await getClientToken(authEndpoint, roomname)
+    _provider = new YSweetProvider(_clientToken.url, roomname, doc, {
+      ...updateProviderParams(providerParams, _clientToken),
+      connect: true,
+    })
+    _provider.addOnFailureHandler(recreateProvider)
+  }
+
+  return {
+    observable,
+    get clientToken() {
+      return _clientToken
+    },
+    destroy() {
+      _provider.destroy()
+    },
+    connectBc() {
+      _provider.connectBc()
+    },
+    disconnectBc() {
+      _provider.disconnectBc()
+    },
+    disconnect() {
+      _provider.disconnect()
+    },
+    connect() {
+      _provider.connect()
+    },
+    addOnFailureHandler(handler: () => void): () => void {
+      return _provider.addOnFailureHandler(handler)
+    },
+    get synced() {
+      return _provider.synced
+    },
+    set synced(state) {
+      _provider.synced = state
+    },
+    get maxBackoffTime() {
+      return _provider.maxBackoffTime
+    },
+    get bcChannel() {
+      return _provider.bcChannel
+    },
+    get url() {
+      return _provider.url
+    },
+    get roomname() {
+      return _provider.roomname
+    },
+    get doc() {
+      return _provider.doc
+    },
+    get awareness() {
+      return _provider.awareness
```

Comment:
things might get weird if the developer holds the awareness across a new provider reconnect. We might need to mock out that class as well, as ugly as that is :/

Reply:
Since we can pass in `awareness` when creating the provider, I'll just use that functionality to keep it referentially stable. See the last commit I pushed there.